design/api.md

```diff
@@ -0,0 +1,219 @@
+Its audience must be `did:<server DID>`.
+
+**Request**:
+```ts
+{
+  code: number,
+  email: string,
+  username: string,
+  did: "did:key:<device>",
+  credentialID?: string,
+}
+```
+
+**Response**:
+
+```ts
+{ ucans: Array<string> }
+```
+
+The response consists of two UCANs:
+- ```json
+  {
+    "iss": "did:key:<account>",
+    "aud": "did:<server DID>",
+    "cap": { "fission:did:key:<account>": { "*": [{}] } }
+  }
+  ```
+- ```json
+  {
+    "iss": "did:<server DID>",
+    "aud": "did:key:<device>",
+    "cap": { "fission:did:key:<account>": { "*": [{}] } }
+  }
+  ```
+
+---
+
+This request registers a full account given an email verification code and delegates the account's access to given DID. In the future more UCANs may be involved in the delgation chain.
+
+
+### GET `/api/v0/account/:did`
+
+---
+
+**Authentication**:
+
+UCAN giving access to the capability `"fission:did:key:<account>": { "account/info": [{}] }` originating in `did:key:<account>`.
+The UCAN's audience needs to be `did:key:<account>`.
+
+**Response**:
+
+```ts
+{
+  email: string,
+  did: string,
+  username: string,
+}
+```
+
+---
+
+This fetches the latest account information.
+
+### GET `/api/v0/account/:username/did`
+
+---
+
+**Authentication**: *Unauthenticated*
+
+**Response**: `{ did: string }`
+
+---
+
+Use this to find out an account's DID. This information may be substituted by DoH & a DNS record in the future.
+
+
+
+
+### PUT `/api/v0/account/:did/username/:username`
+
+---
+
+**Authentication**:
+
+UCAN giving access to the capability `"fission:did:key:<account>": { "account/manage": [{}] }` originating in `did:key:<account>`.
+The UCAN's audience needs to be `did:key:<account>`.
+
+**Response**:
+
+Status 200 OK and `{ success: true }`, if successful.
+
+Status 409 Conflict and `{ success: false }`, if the username is already taken.
+
+---
+
+Change the account's username.
+
+
+### POST `/api/v0/fetch-capabilities/:did`
+
+---
+
+**Authentication**:
+
+```json
+{
+  "iss": "did:<from the path>",
+  "aud": "did:<server DID>",
+  "cap": {}, // Note: Should we introduce a capability? "did:key:zABC": { "capability/fetch": [{}] }?
+  "nnc": "<randomly generated>"
+}
+```
+
+**Request**: `{ ucans: string[] }`
+
+**Response**: `{ ucans: string[] }`
+
+---
+
+This returns a set of UCANs, that, combined with the UCAN from the request body will give the given DID access to new capabilities.
+
+In some cases this can be useful for logging in a new devices. E.g. when passkeys are available and are synced between devices, e.g. it's easy to ask a synced passkey to sign a UCAN giving the `"ucan:*": { "*": [{}] }` capability. However, this UCAN on its own doesn't give the account access to anything yet. It can use this endpoint to fetch the `fission:did:key:<account>` capability UCANs that delegate to the passkey.
+
+In the future this endpoint will be useful for fetching new capabilities from other services that were delegated to a user's account.
```

Comment:
I could a bit confused... is this a request to actually change the username, or a request to get a capability that would allow that? The naming could maybe be clearer. I think I get the use case from the prose below the interface, though.

In REST APIs you tend to see `fetch` get expressed as paths + HTTP method. The wording here lead me to be confised on first glance, as I read it as "do a fetch for capabilities, but with a POST".

REST is intentionally focused on a single resource at a time, which is a motivation for moving to RPC, but a more RESTful design for now:

```http
GET /api/v0/capability/:did/:ability
```
```js
header = {authorization: `bearer ${ucan}`} // or a challenge
```

Example:

```js
const response = await fetch(
  "https://exmaple.com/api/v0/capability/did:key:alice/store/append",
  {authorization: `bearer ${ucan}`}
)

response.payload === [storageUcan1, storageUcan2]
```

When we move to RPC, we probably want a capability request synatx (somthing akin to GNAP). We were joking the other day in Discord about calling it CANI?

Reply:
> "do a fetch for capabilities, but with a POST"

Yeah that's a typo, should be a GET. Sorry!